### PR TITLE
Update docstring for circuit_drawer and QuantumCircuit.draw

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

The docstrings for both the qiskit.visualization.circuit_drawer()
function and the qiskit.circuit.QuantumCircuit.draw() method were quite
stale. With numerous formatting issues and out of date information. This
commit goes through them and updates both the style and contents to
render correctly in sphinx and reflect the current state of the
functions.

### Details and comments

### CODE REVIEW

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -564,65 +564,85 @@ def draw(self, scale=0.7, filename=None, style=None, output=None,
...
+                each gate being placed in its own column.
+            vertical_compression (string): ``high``, ``medium`` or ``low``. It
+                merges the lines generated by the ``text`` output so the
+                drawing will take less vertical room.  Default is ``medium``.
+                Only used by the ``text`` output, will be silently ignored
+                otherwise.
+            idle_wires (bool): Include idle wires (wires with no circuit
+                elements) in output visualization. Default is True.
+            with_layout (bool): Include layout information, with labels on the
+                physical layout.
```
Comment (ajavadia) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
                physical layout. Default is True.
```

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -564,65 +564,85 @@ def draw(self, scale=0.7, filename=None, style=None, output=None,
...
                 In `text`, sets the length of the lines. This useful when the
-                drawing does not fit in the console. If None (default), it will try to
-                guess the console width using `shutil.get_terminal_size()`. However, if
-                running in jupyter, the default line length is set to 80 characters.
-                In `mpl` is the amount of operations before folding. Default is 25.
+                drawing does not fit in the console. If None (default), it will
+                try to guess the console width using ``shutil.
+                get_terminal_size()``. However, if running in jupyter, the
+                default line length is set to 80 characters. In ``mpl`` is the
+                amount of operations before folding. Default is 25.
```
Comment (ajavadia) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
                number of (visual) layers before folding. Default is 25.
```

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -564,65 +564,85 @@ def draw(self, scale=0.7, filename=None, style=None, output=None,
...
+                diagram.
+            * `matplotlib.figure.Figure` (output='mpl')
+                a matplotlib figure object for the circuit diagram.
+            * `str` (output='latex_source')
+                The LaTeX source code for visualizing the circuit diagram.
+            * `TextDrawing` (output='text')
+                A drawing that can be printed as ascii art
         Raises:
             VisualizationError: when an invalid output method is selected
+            ImportError: when the output methods requieres non-installed
```
Comment (ajavadia) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
            ImportError: when the output methods requires non-installed
```

File: qiskit/circuit/quantumcircuit.py [hunk 4]
```diff
@@ -564,65 +564,85 @@ def draw(self, scale=0.7, filename=None, style=None, output=None,
...
 
         Args:
             scale (float): scale of image to draw (shrink if < 1)
             filename (str): file path to save image to
             style (dict or str): dictionary of style or file name of style
-                file. You can refer to the
-                :ref:`Style Dict Doc <style-dict-doc>` for more information
-                on the contents.
+                file. This option is only used by the ``mpl`` output type. If a
+                str ispassed in that is the path to a json file which contains
```
Comment (ajavadia) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
                str is passed in that is the path to a json file which contains
```

File: qiskit/visualization/circuit_visualization.py [hunk 5]
```diff
@@ -58,155 +65,178 @@ def circuit_drawer(circuit,
...
         fold (int): Sets pagination. It can be disabled using -1.
             In `text`, sets the length of the lines. This useful when the
-            drawing does not fit in the console. If None (default), it will try to
-            guess the console width using `shutil.get_terminal_size()`. However, if
-            running in jupyter, the default line length is set to 80 characters.
-            In `mpl` is the amount of operations before folding. Default is 25.
+            drawing does not fit in the console. If None (default), it will try
+            to guess the console width using ``shutil.get_terminal_size()``.
+            However, if running in jupyter, the default line length is set to
+            80 characters. In ``mpl`` is the amount of operations before
```
Comment (ajavadia) [inline qiskit/visualization/circuit_visualization.py]:
```suggestion
            80 characters. In ``mpl`` it is the number of (visual) layers before
```

File: qiskit/visualization/circuit_visualization.py [hunk 6]
```diff
@@ -58,155 +65,178 @@ def circuit_drawer(circuit,
...
+            an option. ``none`` results in each gate being placed in its own
+            column.
+        vertical_compression (string): ``high``, ``medium`` or ``low``. It
+            merges the lines generated by the ``text`` output so the drawing
+            will take less vertical room.  Default is ``medium``. Only used by
+            the ``text`` output, will be silently ignored otherwise.
+        idle_wires (bool): Include idle wires (wires with no circuit elements)
+            in output visualization. Default is True.
+        with_layout (bool): Include layout information, with labels on the
+            physical layout.
```
Comment (ajavadia) [inline qiskit/visualization/circuit_visualization.py]:
```suggestion
            physical layout. Default is True.
```

File: qiskit/visualization/circuit_visualization.py [hunk 7]
```diff
@@ -58,155 +65,178 @@ def circuit_drawer(circuit,
...
+            has an alternative backend set as the default. If the output kwarg
+            is set, that backend will always be used over the default in a user
+            config file.
         interactive (bool): when set true show the circuit in a new window
             (for `mpl` this depends on the matplotlib backend being used
             supporting this). Note when used with either the `text` or the
             `latex_source` output type this has no effect and will be silently
             ignored.
-        line_length (int): Deprecated. See `fold`.
+        line_length (int): Deprecated, see `fold` which superseceds this option.
```
Comment (ajavadia) [inline qiskit/visualization/circuit_visualization.py]:
```suggestion
        line_length (int): Deprecated, see `fold` which supersedss this option.
```

File: qiskit/visualization/circuit_visualization.py [hunk 8]
```diff
@@ -58,155 +65,178 @@ def circuit_drawer(circuit,
...
+            an in-memory representation of the image of the circuit diagram.
+        * `matplotlib.figure.Figure` (output='mpl')
+            a matplotlib figure object for the circuit diagram.
+        * `str` (output='latex_source')
+            The LaTeX source code for visualizing the circuit diagram.
+        * `TextDrawing` (output='text')
+            A drawing that can be printed as ascii art
     Raises:
         VisualizationError: when an invalid output method is selected
         ImportError: when the output methods requieres non-installed libraries.
```
Comment (ajavadia) [inline qiskit/visualization/circuit_visualization.py]:
```suggestion
        ImportError: when the output methods requires non-installed libraries.
```

**Comment by mtreinish** [comment]
> It's probably good to include the extended documentation for the "style" arg in the `circuit.draw()` documentation too.

Sure I can copy it over too. I originally kept it separate because it was pretty long and we could just link to it (like it does there). But I'll make a copy of it for the `draw()` method docstring in one sec

**Pull request merged**